include/unifex/async_resource.hpp

```diff
@@ -0,0 +1,423 @@
+#include <unifex/then.hpp>
+#include <unifex/unstoppable.hpp>
+#include <unifex/v1/async_scope.hpp>
+#include <unifex/when_all.hpp>
+#include <unifex/with_query_value.hpp>
+
+#include <algorithm>
+#include <atomic>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <type_traits>
+#include <utility>
+
+namespace unifex {
+namespace _async_resource {
+/*
+* `ResourceFactory` takes an `async_scope_ref` that refers to the *inner* scope
+* `ResourceFactory(async_scope_ref)` returns a `Resource`
+* the `Resource` is stored next to the *inner* scope in a heap-allocated
+  `_container`
+* `async_resource` takes an *outer* scope ref, returns a move-only
+  `async_resource_ptr<Resource>`
+* the returned `async_resource_ptr<Resource>` is a *unique_ptr*-like object
+* creation of the `Resource`, it's asynchronous destruction and `complete()` of
+  the *inner* scope is spawned on the *outer* scope. Lifetime is controlled by
+  `async_resource_ptr<Resource>`.
+* `async_destroy` is a CPO algorithm that by default does nothing. Customizable
+  through `tag_invoke` to do async destruction if `Resource` provides a
+  `destroy() -> Sender`
+* TBD customizations:
+  - for a range of T, async_destroy(range) is just async_destroy on each T,
+  followed by clearing the range
+  - for an optional<T>, async_destroy(opt<T>) is a no-op on empty
+  - async_destroy(*opt) on non-empty, followed by .reset()
+  - for a unique_ptr<T>, async_destroy is async_destroy(*ptr) followed by
+  .reset(), unless it's nullptr, in which case no-op
+*/
+
+class _container_base {
+protected:
+  enum class state_t : char {
+    ALLOCATED,
+    DESTRUCT_SPAWNED,
+    CONSTRUCTED,
+    DESTROYING,
+  };
+
+  void (*destruct_this_resource_)(_container_base*) noexcept;
+  unifex::any_sender_of<> (*schedule_destruct_)(
+      _container_base*, _container_base*) noexcept;
+  unifex::any_sender_of<> (*destroy_this_resource_)(_container_base*) noexcept;
+  void (*deleter_)(_container_base*) noexcept;
+  // TODO migrate to v2/async_scope before wide adoption
+  UNIFEX_NO_UNIQUE_ADDRESS unifex::manual_lifetime<unifex::v1::async_scope>
+      scope_;
+  _container_base* children_;
+  _container_base* next_;
+  _container_base* parent_;
+  // means destruction should start
+  unifex::async_manual_reset_event destructionEvt_;
+  std::mutex mutex_;
+  std::atomic<state_t> state_;
+
+  unifex::task<void> close_child_scopes() noexcept {
+    auto* adopted = [&]() noexcept {
+      std::lock_guard lock{mutex_};
+      return std::exchange(children_, nullptr);
+    }();
+    {
+      for (auto* child = adopted; child != nullptr; child = child->next_) {
+        co_await child->close_scope();
+      }
+    }
+    {
+      std::lock_guard lock{mutex_};
+      // TODO repeat when children_ != nullptr
+      auto** tail = &children_;
+
+      while (*tail) {
+        tail = &(*tail)->next_;
+      }
+
+      *tail = adopted;
+    }
+  }
+
+  unifex::task<void> destroy_resource() noexcept {
```

Comment:
Since coroutines allocate, this probably shouldn't be `noexcept`.
```suggestion
  unifex::task<void> destroy_resource() {
```